client/internal/dns/server.go

```diff
@@ -127,12 +127,13 @@ func (s *DefaultServer) listen() {
 	// nil check required in unit tests
 	if s.wgInterface != nil && s.wgInterface.IsUserspaceBind() {
 		s.fakeResolverWG.Add(1)
+		s.setListenerStatus(true)
 		go func() {
-			s.setListenerStatus(true)
-			defer s.setListenerStatus(false)
-
 			hookID := s.filterDNSTraffic()
 			s.fakeResolverWG.Wait()
+
+			s.mux.Lock()
+			defer s.mux.Unlock()
```

Comment:
I feel it is not fully thread safe.
With this lock you protect the **listenerIsRunning** variable but you modify that variable from a different thread.
So  you call the s.fakeResolverWG.Done() in 280 line and right after modify the **listenerIsRunning** variable but it does not means that in this thread the s.mux.Lock() will be invoked before you modify.

Reply:
It good to know. On my another task I used it in wrong way I think.

Reply:
In this case I recommend you to use channel withe a lock instead of wg.  With it you can avoid the weakness (negative write) of wg. The code style will be ugly but by the way :)

Reply:
When I get a negative number, it means we have something wrong with calls, and it helps to catch bugs.
Anyway - the workgroup was removed :)

Reply:
You definitely found the greatest solution! I like it :)